Stack evolves, use lucky egg to maximize XP

svombergen opened the pull request:
### Prerequisites
Make sure you remove the pokemons you want to keep for evolving from the ObligatoryTransferList and add them to the IgnoredPokemonList

**Changes made: 	**

 1. Facilitate evolve stacking, configurable with 2 variables: 
- evolveBeforeTransfer - list of pokemon to save up candies and pokemon for evolve stacking
- evolveStackLimit - threshold to pop lucky egg and evolve all saved up evolves at once

 2. Made logging spam of walkingLoop cycle dependent on displayKeepalive setting.

** Changes detail**

- Added 2 variables in settings and template file, evolveBeforeTransfer and evolveStackLimit
- Added EvolvePokemon task to handle Evolve stacking logic.
- Added call to task EvolvePokemon in Bot.kt ProfileLoop
- Added 'if (settings.displayKeepalive)' statement to logging message in WalkingLoop of Walk.kt to avoid spam in console.

zhoushifeng8 commented:
when can use this evolve function?

svombergen commented:
When jabbink decides to merge it.

svombergen commented:
I expect the release logic could use a good refactoring, after adding evolves. The logic for releasing will be divided and intertwined on 3 places.

jabbink commented on src/main/kotlin/ink/abb/pogo/scraper/tasks/EvolvePokemon.kt:
```diff
@@ -0,0 +1,79 @@
+                it.value.sortedByDescending { it.cp }.forEach {
+                    maxPossibleEvolves--
+                    if (maxPossibleEvolves < 0) {
+                        releasePokemon(it,"Not enough candy to save for evolve",ctx)
```
releasing a pokemon increases your candy count so you could update the maxPossibleEvolves again :smile:

svombergen replied:
oh you are right, my lazy code will just check next profileloop rotation :smile:

jabbink commented:
Can you please use the default IntelliJ autoformat eg `if(settings` becomes `if (settings`

Also how does this work with transferring pokemon based on the cp/iv thresholds and maxPokemonAmount/keepPokemonAmount?

jabbink commented on src/main/kotlin/ink/abb/pogo/scraper/tasks/EvolvePokemon.kt:
```diff
@@ -0,0 +1,79 @@
+            if(settings.evolveBeforeTransfer.contains(it.key)){
+                // Get pokemonFamily through this val (can get this somewhere else?):
+                val lastPokemonInGroup = it.value.last()
+                var maxPossibleEvolves = bot.api.inventories.candyjar.getCandies(lastPokemonInGroup.pokemonFamily) / lastPokemonInGroup.candiesToEvolve
```
```kotlin

                val pokemonMeta = PokemonMetaRegistry.getMeta(it.key)
                var maxPossibleEvolves = bot.api.inventories.candyjar.getCandies(pokemonMeta.family) / pokemonMeta.candyToEvolve
```

svombergen replied:
ah thank you, will update

svombergen commented:
Prerequisite is putting the pokemons you want to evolve into the ignoredPokemon list (and remove from obligatory). The lists are getting pretty redundant, so evolve and release tasks could use a good refactoring.

jabbink commented:
>so evolve and release tasks could use a good refactoring.

agreed

jabbink commented:
Will test now

devinsba commented on src/main/kotlin/ink/abb/pogo/scraper/tasks/EvolvePokemon.kt:
```diff
@@ -0,0 +1,80 @@
+                // Add the minimum value, depending on which is the bottleneck, amount of candy, or pokemon of this type in pokebank:
+                countEvolveStack +=  Math.min(maxPossibleEvolves,it.value.count())
+                // Release pokemon we wont be able to evolve anyway, because of lack of candy:
+                it.value.sortedByDescending { it.cp }.forEach {
```
Shouldn't this use IV if their settings have IV selected as their sort?

jabbink commented on src/main/kotlin/ink/abb/pogo/scraper/tasks/EvolvePokemon.kt:
```diff
@@ -0,0 +1,80 @@
+        if (countEvolveStack >= settings.evolveStackLimit) {
+            val startingXP = ctx.api.playerProfile.stats.experience
+            Log.yellow("Using Lucky Egg before evolving stack of $countEvolveStack pokemons")
+            var resultLuckyEgg = ctx.api.inventories.itemBag.useLuckyEgg()
```
please use `ctx.api.cachedInventories` and wrap in try/catch because it may fail

jabbink commented:
Looks good, please see line comments and auto indentation issue.

svombergen commented:
Just merged develop branch and:
- Added try catch around lucky egg usage, also changed to cachedInventory
- Added sorting by Iv when releasing
- Corrected indentation

jabbink commented:
>Prerequisite is putting the pokemons you want to evolve into the ignoredPokemon list (and remove from obligatory).

Maybe to prevent confusion when this gets merged, change the shouldTransfer function(s?) to early return when the pokemon is in the evolveBeforeTransfer list

jabbink commented:
also about overhaul #860

svombergen commented:
Agreed, will do some refactoring to prevent confusion later today.

yuanweilin commented:
add the 
"evolve_before_transfer" : [ "DODUO", "RATTATA", "CATERPIE", "PIDGEY", "ZUBAT", "GOLDEEN", "MAGMAR", "PINSIR", "GOLBAT", "WEEDLE" ],
 "evolve_stack_limit" : 100,
in my setting
when i run java -jar PokemonGoBot-0.5.2.jar 
It will be deleted

svombergen commented:
because it is not merged into 0.5.2, just a little more patience :)

yuanweilin commented:
lol,  I'm so sad

svombergen commented:
Alright committed shouldrelease logic to avoid confusion. Will truely have a look at refactoring all logic around evolving and releasing when i get the time.

gise88 commented:
I know.. I am in late because this was already merged but I am going to ask if you can watch at this: https://github.com/jabbink/PokemonGoBot/issues/1040

Thanks anyway! :smiley:

Sieberkev commented:
Halfway the evolving (50/100 pokes) I got:

```
14 aug 13:58:59 [default: ProfileLoop] - Error running loop ProfileLoop!
com.pokegoapi.exceptions.AsyncPokemonGoException: Unknown exception occurred.
        at com.pokegoapi.util.AsyncHelper.toBlocking(AsyncHelper.java:45)
        at com.pokegoapi.main.RequestHandler.sendServerRequests(RequestHandler.java:155)
        at com.pokegoapi.api.pokemon.Pokemon.evolve(Pokemon.java:231)
        at ink.abb.pogo.scraper.tasks.EvolvePokemon.run(EvolvePokemon.kt:69)
        at ink.abb.pogo.scraper.Bot.task(Bot.kt:254)
        at ink.abb.pogo.scraper.Bot$start$3.invoke(Bot.kt:142)
        at ink.abb.pogo.scraper.Bot$start$3.invoke(Bot.kt:35)
        at ink.abb.pogo.scraper.Bot$runLoop$1.invoke(Bot.kt:202)
        at ink.abb.pogo.scraper.Bot$runLoop$1.invoke(Bot.kt:35)
        at kotlin.concurrent.ThreadsKt$thread$thread$1.run(Thread.kt:18)
Caused by: java.lang.RuntimeException: java.util.concurrent.ExecutionException: com.google.protobuf.InvalidProtocolBufferException: Contents of buffer are null
        at rx.exceptions.Exceptions.propagate(Exceptions.java:58)
        at rx.observables.BlockingObservable.blockForSingle(BlockingObservable.java:465)
        at rx.observables.BlockingObservable.first(BlockingObservable.java:168)
        at com.pokegoapi.util.AsyncHelper.toBlocking(AsyncHelper.java:37)
        ... 9 more
Caused by: java.util.concurrent.ExecutionException: com.google.protobuf.InvalidProtocolBufferException: Contents of buffer are null
        at com.pokegoapi.main.RequestHandler$1.get(RequestHandler.java:109)
        at com.pokegoapi.main.RequestHandler$1.get(RequestHandler.java:86)
        at rx.internal.operators.OnSubscribeToObservableFuture$ToObservableFuture.call(OnSubscribeToObservableFuture.java:74)
        at rx.internal.operators.OnSubscribeToObservableFuture$ToObservableFuture.call(OnSubscribeToObservableFuture.java:43)
        at rx.internal.operators.OnSubscribeLift.call(OnSubscribeLift.java:48)
        at rx.internal.operators.OnSubscribeLift.call(OnSubscribeLift.java:30)
        at rx.internal.operators.OnSubscribeLift.call(OnSubscribeLift.java:48)
        at rx.internal.operators.OnSubscribeLift.call(OnSubscribeLift.java:30)
        at rx.Observable.subscribe(Observable.java:9957)
        at rx.Observable.subscribe(Observable.java:9924)
        at rx.observables.BlockingObservable.blockForSingle(BlockingObservable.java:445)
        ... 11 more
Caused by: com.google.protobuf.InvalidProtocolBufferException: Contents of buffer are null
        at com.pokegoapi.main.ServerRequest.getData(ServerRequest.java:78)
        at com.pokegoapi.main.RequestHandler.run(RequestHandler.java:296)
        at java.lang.Thread.run(Thread.java:745)

```

Maybe the evolving is too fast?

johnnyslt commented:
Same thing for me, while using the bot from source:

`      14 Aug 22:26:31 [default: ProfileLoop] - Going to transfer RATICATE with CP 749 and IV 53%; reason: Evolved for XP only
       14 Aug 22:26:32 [default: ProfileLoop] - Evolving PIDGEY before transfer
       14 Aug 22:26:32 [default: ProfileLoop] - Error running loop ProfileLoop!
com.pokegoapi.exceptions.AsyncPokemonGoException: Unknown exception occurred.
        at com.pokegoapi.util.AsyncHelper.toBlocking(AsyncHelper.java:45)
        at com.pokegoapi.main.RequestHandler.sendServerRequests(RequestHandler.java:155)
        at com.pokegoapi.api.pokemon.Pokemon.evolve(Pokemon.java:231)
        at ink.abb.pogo.scraper.tasks.EvolvePokemon.run(EvolvePokemon.kt:69)
        at ink.abb.pogo.scraper.Bot.task(Bot.kt:254)
        at ink.abb.pogo.scraper.Bot$start$3.invoke(Bot.kt:142)
        at ink.abb.pogo.scraper.Bot$start$3.invoke(Bot.kt:35)
        at ink.abb.pogo.scraper.Bot$runLoop$1.invoke(Bot.kt:202)
        at ink.abb.pogo.scraper.Bot$runLoop$1.invoke(Bot.kt:35)
        at kotlin.concurrent.ThreadsKt$thread$thread$1.run(Thread.kt:18)
Caused by: java.lang.RuntimeException: java.util.concurrent.ExecutionException: com.google.protobuf.InvalidProtocolBufferException: Contents of buffer are null
        at rx.exceptions.Exceptions.propagate(Exceptions.java:58)
        at rx.observables.BlockingObservable.blockForSingle(BlockingObservable.java:465)
        at rx.observables.BlockingObservable.first(BlockingObservable.java:168)
        at com.pokegoapi.util.AsyncHelper.toBlocking(AsyncHelper.java:37)
        ... 9 more
Caused by: java.util.concurrent.ExecutionException: com.google.protobuf.InvalidProtocolBufferException: Contents of buffer are null
        at com.pokegoapi.main.RequestHandler$1.get(RequestHandler.java:109)
        at com.pokegoapi.main.RequestHandler$1.get(RequestHandler.java:86)
        at rx.internal.operators.OnSubscribeToObservableFuture$ToObservableFuture.call(OnSubscribeToObservableFuture.java:74)
        at rx.internal.operators.OnSubscribeToObservableFuture$ToObservableFuture.call(OnSubscribeToObservableFuture.java:43)
        at rx.internal.operators.OnSubscribeLift.call(OnSubscribeLift.java:48)
        at rx.internal.operators.OnSubscribeLift.call(OnSubscribeLift.java:30)
        at rx.internal.operators.OnSubscribeLift.call(OnSubscribeLift.java:48)
        at rx.internal.operators.OnSubscribeLift.call(OnSubscribeLift.java:30)
        at rx.Observable.subscribe(Observable.java:9957)
        at rx.Observable.subscribe(Observable.java:9924)
        at rx.observables.BlockingObservable.blockForSingle(BlockingObservable.java:445)
        ... 11 more
Caused by: com.google.protobuf.InvalidProtocolBufferException: Contents of buffer are null
        at com.pokegoapi.main.ServerRequest.getData(ServerRequest.java:78)
        at com.pokegoapi.main.RequestHandler.run(RequestHandler.java:296)
        at java.lang.Thread.run(Thread.java:745)
`

johnnyslt commented:
I get the same thing:

>       14 Aug 22:26:31 [default: ProfileLoop] - Going to transfer RATICATE with CP 749 and IV 53%; reason: Evolved for XP only
       14 Aug 22:26:32 [default: ProfileLoop] - Evolving PIDGEY before transfer
       14 Aug 22:26:32 [default: ProfileLoop] - Error running loop ProfileLoop!
com.pokegoapi.exceptions.AsyncPokemonGoException: Unknown exception occurred.
        at com.pokegoapi.util.AsyncHelper.toBlocking(AsyncHelper.java:45)
        at com.pokegoapi.main.RequestHandler.sendServerRequests(RequestHandler.java:155)
        at com.pokegoapi.api.pokemon.Pokemon.evolve(Pokemon.java:231)
        at ink.abb.pogo.scraper.tasks.EvolvePokemon.run(EvolvePokemon.kt:69)
        at ink.abb.pogo.scraper.Bot.task(Bot.kt:254)
        at ink.abb.pogo.scraper.Bot$start$3.invoke(Bot.kt:142)
        at ink.abb.pogo.scraper.Bot$start$3.invoke(Bot.kt:35)
        at ink.abb.pogo.scraper.Bot$runLoop$1.invoke(Bot.kt:202)
        at ink.abb.pogo.scraper.Bot$runLoop$1.invoke(Bot.kt:35)
        at kotlin.concurrent.ThreadsKt$thread$thread$1.run(Thread.kt:18)
Caused by: java.lang.RuntimeException: java.util.concurrent.ExecutionException: com.google.protobuf.InvalidProtocolBufferException: Contents of buffer are null
        at rx.exceptions.Exceptions.propagate(Exceptions.java:58)
        at rx.observables.BlockingObservable.blockForSingle(BlockingObservable.java:465)
        at rx.observables.BlockingObservable.first(BlockingObservable.java:168)
        at com.pokegoapi.util.AsyncHelper.toBlocking(AsyncHelper.java:37)
        ... 9 more
Caused by: java.util.concurrent.ExecutionException: com.google.protobuf.InvalidProtocolBufferException: Contents of buffer are null
        at com.pokegoapi.main.RequestHandler$1.get(RequestHandler.java:109)
        at com.pokegoapi.main.RequestHandler$1.get(RequestHandler.java:86)
        at rx.internal.operators.OnSubscribeToObservableFuture$ToObservableFuture.call(OnSubscribeToObservableFuture.java:74)
        at rx.internal.operators.OnSubscribeToObservableFuture$ToObservableFuture.call(OnSubscribeToObservableFuture.java:43)
        at rx.internal.operators.OnSubscribeLift.call(OnSubscribeLift.java:48)
        at rx.internal.operators.OnSubscribeLift.call(OnSubscribeLift.java:30)
        at rx.internal.operators.OnSubscribeLift.call(OnSubscribeLift.java:48)
        at rx.internal.operators.OnSubscribeLift.call(OnSubscribeLift.java:30)
        at rx.Observable.subscribe(Observable.java:9957)
        at rx.Observable.subscribe(Observable.java:9924)
        at rx.observables.BlockingObservable.blockForSingle(BlockingObservable.java:445)
        ... 11 more
Caused by: com.google.protobuf.InvalidProtocolBufferException: Contents of buffer are null
        at com.pokegoapi.main.ServerRequest.getData(ServerRequest.java:78)
        at com.pokegoapi.main.RequestHandler.run(RequestHandler.java:296)
        at java.lang.Thread.run(Thread.java:745)
